JDK-8263558: Possible NULL dereference in fast path arena free if ZapResourceArea is true

tstuefe opened the pull request:
Sonarcloud reports a possible NULL dereference when zapping the to-be-freed area in fast-path arena free. Possible call stack for this to happen starts in Node::destruct(PhaseValues* phase).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263558](https://bugs.openjdk.java.net/browse/JDK-8263558): Possible NULL dereference in fast path arena free if ZapResourceArea is true


### Reviewers
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to ceaef852bc00274c46e97d27735c651588254e07


### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/2995/head:pull/2995`
`$ git checkout pull/2995`

To update a local copy of the PR:
`$ git checkout pull/2995`
`$ git pull https://git.openjdk.java.net/jdk pull/2995/head`

kimbarrett commented on src/hotspot/share/memory/arena.hpp:
```diff
@@ -185,7 +185,9 @@ class Arena : public CHeapObj<mtNone> {
   bool Afree(void *ptr, size_t size) {
 #ifdef ASSERT
-    if (ZapResourceArea) memset(ptr, badResourceValue, size); // zap freed memory
+    if (ZapResourceArea && ptr != NULL) {
```
The pointer arithmetic on line 193 is UB if ptr is nullptr (though it's unlikely anyone checks), so either Afree has `ptr != nullptr` as an (unstated) precondition or there are potentially more problems than the memset.

tstuefe commented:
Thanks Kim. I changed the fix to fix the one caller I was sure about passing NULL, and assert in AFree instead. 

I removed the Trivial mark. Before pushing I will put this through tests to check if it triggers. I believe even though the pointer arithmetic below was UB with NULL, the effect would have in general been benign (just refusing to free anything).

tstuefe commented:
> _Mailing list message from [Kim Barrett](mailto:kim.barrett@oracle.com) on [hotspot-runtime-dev](mailto:hotspot-runtime-dev@openjdk.java.net):_
> 
> Looks good.

Thanks Kim.

coleenp reviewed (approved):
Looks good.

tstuefe commented:
Thanks @coleenp and @kimbarrett . However, as I feared, with the assert alone I now see it firing both in our nightlies at SAP and in the GAs in compiler tests. 

Which is cool in a way since this is the first time SonarCloud reported anything demonstrably real.:) I'll take a look at the crashes and modify the patch.

tstuefe commented:
I changed the patch to provide the same semantics on passed old ptr == NULL as standard `free(3)` and `realloc(3)` have. Which is to ignore it (free) or to revert to malloc (realloc).

The culprit in this case - calling realloc with a NULL pointer - was aot, `ImplicitExceptionTable::append( uint exec_off, uint cont_off )`. 

I considered fixing the caller, but I am used to the C-runtime semantics in free and realloc, and it looks like others are too; so this is the least surprising behavior for a realloc-like function.

Also, this fixes the subtle bug where, when passing NULL to Arealloc, we would return "false" to indicate that we have a lossfull realloc. The only case I can see where the return value was actually used was in `SymbolTable::delete_symbol()` which would print something about leaked symbols in that case. I did not investigate whether this has any practical relevance.

tstuefe commented:
Coleen, Kim, are you fine with this latest version? Thanks!

kimbarrett commented on src/hotspot/share/memory/arena.cpp:
```diff
@@ -366,6 +366,10 @@ void* Arena::grow(size_t x, AllocFailType alloc_failmode) {
 // Reallocate storage in Arena.
 void *Arena::Arealloc(void* old_ptr, size_t old_size, size_t new_size, AllocFailType alloc_failmode) {
   if (new_size == 0) return NULL;
```
[pre-existing] Isn't this a leak?  Probably just dropping old_ptr.

tstuefe replied:
Yes, it is a leak, at least if we run with UseMallocOnly. I think this, and Amalloc, should behave consistently with os::malloc here, which is to return a small non-null allocation for size=0. Basically, if size==0 size=1.

kimbarrett commented on src/hotspot/share/memory/arena.cpp:
```diff
@@ -366,6 +366,10 @@ void* Arena::grow(size_t x, AllocFailType alloc_failmode) {
 // Reallocate storage in Arena.
 void *Arena::Arealloc(void* old_ptr, size_t old_size, size_t new_size, AllocFailType alloc_failmode) {
   if (new_size == 0) return NULL;
+  if (old_ptr == NULL) {
```
If we're being consistent with realloc, probably should test this first, then for new_size == 0 (-> Afree, see above), then the rest.

tstuefe replied:
I would prefer to handle it as I wrote above.

(side note, I would love to rewrite the arena coding; either that or throw it away and reuse the existing Metaspace implementation instead for Arenas/RA. Metaspaces are Arenas too, they do the same thing, but Metaspace is a lot more evolved, and I am biased toward its code quality oc :)

tstuefe replied:
I fixed this by calling Afree in case new_size is 0. Note that Afree is a noop if the pointer is NULL.

kimbarrett reviewed (approved):
Looks good.

tstuefe commented:
Thanks Coleen & Kim!

/integrate
